### PR TITLE
[plugins] Allocate shell handlers once per build.

### PR DESCRIPTION
**Pull request by ddunbar**
- This is more efficient, while also allowing extensions to easily tie their
   lifetime to the lifetime of the build.

 - This also allows the buildsystem to avoid a lock on the common lookup path.

### CODE REVIEW

**Comment by ddunbar** [comment]
@swift-ci please smoke test

File: lib/BuildSystem/BuildSystem.cpp [hunk 1]
```diff
@@ -208,6 +208,9 @@ class BuildSystemImpl : public BuildSystemCommandInterface {
 
   /// Flag indicating if the build has been cancelled.
   std::atomic<bool> isCancelled_{ false };
+
+  /// Cache of instantiated shell command handlers.
+  llvm::StringMap<std::unique_ptr<ShellCommandHandler>> shellHandlers;
```
Comment (dmbryson) [inline lib/BuildSystem/BuildSystem.cpp]: Shouldn't we be clearing this at the start of build()?   As it currently stands we will reuse these across multiple builds if the engine is cached.

Comment (ddunbar) [reply lib/BuildSystem/BuildSystem.cpp]: Oops, yes, indeed.

**Comment by dmbryson** [comment]
@swift-ci please smoke test

**Comment by dmbryson** [comment]
@swift-ci please smoke test osx

**Comment by dmbryson** [comment]
@swift-ci please smoke test osx

**Comment by dmbryson** [comment]
@swift-ci please smoke test

**Pull request merged**